airflow/utils/task_group.py

```diff
@@ -0,0 +1,233 @@
+        """
+        for task in self:
+            if not any(child.is_in_task_group(self.group_ids)
+                       for child in task.get_direct_relatives(upstream=False)):
+                yield task
+
+    def __rshift__(self, other):
+        """
+        Implements Self >> Other == self.set_downstream(other)
+        """
+        self.set_downstream(other)
+        return other
+
+    def __lshift__(self, other):
+        """
+        Implements Self << Other == self.set_upstream(other)
+        """
+        self.set_upstream(other)
+        return other
+
+    def __rrshift__(self, other):
+        """
+        Called for Operator >> [Operator] because list don't have
+        __rshift__ operators.
+        """
+        self.__lshift__(other)
+        return self
+
+    def __rlshift__(self, other):
+        """
+        Called for Operator << [Operator] because list don't have
+        __lshift__ operators.
+        """
+        self.__rshift__(other)
+        return self
+
+    @classmethod
+    def build_task_group(cls, tasks):
+        """
+        Put tasks into TaskGroup.
+        """
+        root = TaskGroup.create_root()
+        for task in tasks:
+            current = root
+            for label in task.task_group_ids:
+                if label not in current.children:
+                    child_group = TaskGroup(group_id=label, parent_group=current)
+                else:
+                    child_group = current.children[label]
+                current = child_group
+            current.add(task)
+        return root
+
+
+class TaskGroupContext:
+    """
+    TaskGroup context is used to keep the current TaskGroup when TaskGroup is used as ContextManager.
+    """
+
+    _context_managed_task_group: Optional[TaskGroup] = None
+    _previous_context_managed_task_groups: List[TaskGroup] = []
+
+    @classmethod
+    def push_context_managed_task_group(cls, task_group: TaskGroup):
+        """
+        Push a TaskGroup into the list of managed TaskGroups.
+        """
+        if cls._context_managed_task_group:
+            cls._previous_context_managed_task_groups.append(cls._context_managed_task_group)
+        cls._context_managed_task_group = task_group
+
+    @classmethod
+    def pop_context_managed_task_group(cls) -> Optional[TaskGroup]:
+        """
+        Pops the last TaskGroup from the list of manged TaskGroups and update the current TaskGroup.
+        """
+        old_task_group = cls._context_managed_task_group
+        if cls._previous_context_managed_task_groups:
+            cls._context_managed_task_group = cls._previous_context_managed_task_groups.pop()
+        else:
+            cls._context_managed_task_group = None
+        return old_task_group
+
+    @classmethod
+    def get_current_task_group(cls) -> Optional[TaskGroup]:
+        """
+        Get the current TaskGroup.
+        """
+        if not cls._context_managed_task_group:
+            return TaskGroup.create_root()
```

Comment:
should a user expect `_context_managed_task_group` to always point to current task group? if so, should we set `_context_managed_task_group` to this newly created task group before returning?

Reply:
This isn't an issue at the moment because only the `task_group_ids` is important, but you are right this is a potential booby trap for future maintenance. I'll fix this.

Reply:
Made the behaviour like this: If there's currently a DAG but no TaskGroup, return the root TaskGroup of the dag.